src/Shared/CommunicationsUtilities.cs

```diff
@@ -330,14 +330,17 @@ internal static long GetHostHandshake(HandshakeOptions nodeType)
 
             Trace("MSBUILDNODEHANDSHAKESALT=\"{0}\", msbuildDirectory=\"{1}\", nodeType={2}, FileVersionHash={3}", salt, toolsDirectory, nodeType, FileVersionHash);
 
-            //FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
-            //nodeType (4 bits) is shifted just after the FileVersionHash
-            //nodeHandshakeSalt (32 bits) is shifted just after hostContext
-            //the most significant byte (leftmost 8 bits) will get zero'd out to avoid connecting to older builds.
-            //| masked out | nodeHandshakeSalt | hostContext |              fileVersionHash             | SessionID
-            //  0000 0000     0000 0000 0000        0000        0000 0000 0000 0000 0000 0000 0000 0000   0000 0000
-            long baseHandshake = ((long)nodeHandshakeSalt << 44) | ((long)nodeType << 40) | ((long)FileVersionHash << 8);
-            return GenerateHostHandshakeFromBase(baseHandshake);
+            // FileVersionHash (32 bits) is shifted 8 bits to avoid session ID collision
+            // HandshakeOptions (4 bits) is shifted just after the FileVersionHash
+            // nodeHandshakeSalt (32 bits) is xored with the result
+            //  HandshakeOptions |             fileVersionHash           | SessionID
+            //       0000        0000 0000 0000 0000 0000 0000 0000 0000  0000 0000
+            unchecked
+            {
+                ulong baseHandshake = ((ulong)nodeType << 40) | ((ulong)FileVersionHash << 8);
+                baseHandshake ^= (ulong)nodeHandshakeSalt << 32 + nodeHandshakeSalt; // shift and add to give real values for the whole long
```

Comment:
`+` has higher precedence than `<<` so this is not doing the right thing.

Also:

The reason you're xor'ing the result with the salt is that you can use the full 32 bits of the salt now, is that correct? It introduces potential collisions between the salt the other fields, however. For example:

`nodeType = NodeType.None, FileVersionHash = 0, nodeHandshakeSalt = 256`

and

`nodeType = NodeType.TaskHost, FileVersionHash = 1, nodeHandshakeSalt = 0`

both yield the same baseHandshake of 1099511628032 so this has the same problem as just encapsulating all these integers in a struct and computing its hash. I'd expect this to be undesired and would rather be open to more collisions within the handshake salt but keeping the guarantee that a task host will never be mistaken for a non-task host. What do you think?

Reply:
I should have checked that, but I'd thought all bitwise operators were higher precedence than arithmetic ones. Thanks!

I may have misunderstood what we wanted to do, but I thought that was actually what we were trying for. The salt can still be used to prevent nodes from communicating as long as you don't choose one of the few salts that would cause a collision, and this means that if we have a mistake in the hash computation, we can ask users to set the salt to a specific value, and their problems may magically disappear pending a real fix.

Reply:
Thinking about this with Monday Brain, I think I'm back on "salt only the non-options bits".

I was thinking that it'd be nice to be able to drop particular bits out of the hash computation. But that doesn't make sense in practice, because the environment variable will be inherited to child processes. So suppose we decided to mask `TaskHost`--we'd need to have one salt to set it up right in the scheduler process, and another to set it up in the TaskHost workers, or we'd get the same kind of cross-node-type communication we're getting with the current buggy code.

Sorry for the waffling here @Forgind.